sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -42,10 +42,10 @@ import org.apache.spark.sql.types._
  */
 object ConstantFolding extends Rule[LogicalPlan] {
 
-  private def hasNoSideEffect(e: Expression): Boolean = e match {
+  private def cannotThrowException(e: Expression): Boolean = e match {
     case _: Attribute => true
     case _: Literal => true
-    case _: NoSideEffect => e.children.forall(hasNoSideEffect)
+    case _: NoThrow => e.children.forall(cannotThrowException)
```

Comment:
Since the naming is clear now, I'd suggest we add `if e.determinisitc` here, since `NoThrow` doesn't indicate that it's deterministic.

Reply:
👌